sdks/java/io/cdap/src/main/java/org/apache/beam/sdk/io/cdap/CdapIO.java

```
@@ -0,0 +1,235 @@
+
+    @Experimental(Experimental.Kind.PORTABILITY)
+    @AutoValue.Builder
+    abstract static class Builder<K, V> {
+
+      abstract Builder<K, V> setPluginConfig(PluginConfig config);
+
+      abstract Builder<K, V> setCdapPlugin(Plugin plugin);
+
+      abstract Builder<K, V> setKeyClass(Class<K> keyClass);
+
+      abstract Builder<K, V> setValueClass(Class<V> valueClass);
+
+      abstract Read<K, V> build();
+    }
+
+    public Read<K, V> withCdapPlugin(Plugin plugin) {
+      checkArgument(plugin != null, "Cdap plugin can not be null");
+      return toBuilder().setCdapPlugin(plugin).build();
+    }
+
+    public Read<K, V> withCdapPluginClass(Class<?> cdapPluginClass) {
+      checkArgument(cdapPluginClass != null, "Cdap plugin class can not be null");
+      Plugin plugin = MappingUtils.getPluginByClass(cdapPluginClass);
+      checkArgument(plugin != null, "Can not instantiate Plugin object from provided plugin class");
+      return toBuilder().setCdapPlugin(plugin).build();
+    }
+
+    public Read<K, V> withPluginConfig(PluginConfig pluginConfig) {
+      checkArgument(pluginConfig != null, "Plugin config can not be null");
+      return toBuilder().setPluginConfig(pluginConfig).build();
+    }
+
+    public Read<K, V> withKeyClass(Class<K> keyClass) {
+      checkArgument(keyClass != null, "Key class can not be null");
+      return toBuilder().setKeyClass(keyClass).build();
+    }
+
+    public Read<K, V> withValueClass(Class<V> valueClass) {
+      checkArgument(valueClass != null, "Value class can not be null");
+      return toBuilder().setValueClass(valueClass).build();
+    }
+
+    @Override
+    public PCollection<KV<K, V>> expand(PBegin input) {
+      validateTransform();
+
+      getCdapPlugin()
+          .withConfig(getPluginConfig())
+          .withHadoopConfiguration(getKeyClass(), getValueClass())
+          .prepareRun();
+
+      if (getCdapPlugin().isUnbounded()) {
+        // TODO: implement SparkReceiverIO.<~>read()
```

This should probably throw an exception for now, to prevent confusion

---

Done